x-pack/plugins/actions/server/builtin_action_types/servicenow/utils.ts

```diff
@@ -69,3 +82,129 @@ export const throwIfSubActionIsNotSupported = ({
     throw new Error(errorMessage);
   }
 };
+
+export interface GetAccessTokenAndAxiosInstanceOpts {
+  connectorId: string;
+  logger: Logger;
+  configurationUtilities: ActionsConfigurationUtilities;
+  credentials: ExternalServiceCredentials;
+  snServiceUrl: string;
+  connectorTokenClient: ConnectorTokenClientContract;
+}
+
+export const getAxiosInstance = ({
+  connectorId,
+  logger,
+  configurationUtilities,
+  credentials,
+  snServiceUrl,
+  connectorTokenClient,
+}: GetAccessTokenAndAxiosInstanceOpts): AxiosInstance => {
+  const { config, secrets } = credentials;
+  const { isOAuth } = config as ServiceNowPublicConfigurationType;
+  const { username, password } = secrets as ServiceNowSecretConfigurationType;
+
+  let axiosInstance;
+
+  if (!isOAuth && username && password) {
+    axiosInstance = axios.create({
+      auth: { username, password },
+    });
+  } else {
+    axiosInstance = axios.create();
+    axiosInstance.interceptors.request.use(
+      async (axiosConfig: AxiosRequestConfig) => {
+        const accessToken = await getAccessToken({
+          connectorId,
+          logger,
+          configurationUtilities,
+          credentials: {
+            config: config as ServiceNowPublicConfigurationType,
+            secrets,
+          },
+          snServiceUrl,
+          connectorTokenClient,
+        });
+        axiosConfig.headers.Authorization = accessToken;
+        return axiosConfig;
+      },
+      (error) => {
+        Promise.reject(error);
+      }
+    );
+  }
+
+  return axiosInstance;
+};
+
+export const getAccessToken = async ({
+  connectorId,
+  logger,
+  configurationUtilities,
+  credentials,
+  snServiceUrl,
+  connectorTokenClient,
+}: GetAccessTokenAndAxiosInstanceOpts) => {
+  const { isOAuth, clientId, jwtKeyId, userIdentifierValue } =
+    credentials.config as ServiceNowPublicConfigurationType;
+  const { clientSecret, privateKey, privateKeyPassword } =
+    credentials.secrets as ServiceNowSecretConfigurationType;
+
+  let accessToken: string;
+
+  // Check if there is a token stored for this connector
+  const { connectorToken, hasErrors } = await connectorTokenClient.get({ connectorId });
+
+  if (connectorToken === null || Date.parse(connectorToken.expiresAt) <= Date.now()) {
```

Comment:
Are we sure that `connectorToken.expiresAt` will always be valid? If not `Date.parse` will return `NaN` and  `NaN <= Date.now()` always equals `false`.

Reply:
That's a good point! I am creating a follow up issue to harden the connector token workflow as Mike pointed out another edge case to watch out for so I will add this to the follow up issue.